### ISSUE TITLE
How to use feign client between my services with csrf protection ?

### ISSUE THREAD

**Issue by maxiplay** [spring-cloud/spring-cloud-netflix#823]
I have 403 when I attempt to contact my services with feign. How to manage it ? I want csrf protection.

```
Caused by: feign.FeignException: status 403 reading CharacterApi#createNewCharacterList(String); content:
{"timestamp":1455221787486,"status":403,"error":"Forbidden","message":"Expected CSRF token not found. Has your session expired?","path":"/characterList/max3"}
	at feign.FeignException.errorStatus(FeignException.java:62) ~[feign-core-8.12.1.jar:8.12.1]
	at feign.codec.ErrorDecoder$Default.decode(ErrorDecoder.java:91) ~[feign-core-8.12.1.jar:8.12.1]
	at feign.SynchronousMethodHandler.executeAndDecode(SynchronousMethodHandler.java:126) ~[feign-core-8.12.1.jar:8.12.1]
	at feign.SynchronousMethodHandler.invoke(SynchronousMethodHandler.java:74) ~[feign-core-8.12.1.jar:8.12.1]
	at feign.hystrix.HystrixInvocationHandler$1.run(HystrixInvocationHandler.java:54) ~[feign-hystrix-8.12.1.jar:8.12.1]
	at com.netflix.hystrix.HystrixCommand$1.call(HystrixCommand.java:294) ~[hystrix-core-1.4.21.jar:1.4.21]
```

**Comment by dsyer** [comment]
If you have Spring Security on the server you can ask it to send you a cookie with the CSRF token, or you can set up an endpoint to read the session cookie and CSRF token from. You are going to have to send the session cookie as well, so be prepared for some header schlepping and some state management work.

 I can't really see why a Feign client needs to send a CSRF token though. Do you have a scenario where you are worried about a specific threat?

**Comment by maxiplay** [comment]
Hi, thanks for reply.

Yes I have spring security on my servers. Each server manage its own security concern but authentication is on gateway with zull and shared redis.

Each micro services is exposed through the gateway so I need csrf protection. 
The use case is tha

**Comment by dsyer** [comment]
So your backend resources are used both by a browser (XHR) and also other services (Feign)? Could you not split them into two populations to be used by one or the other (it's easier to secure endpoints based on path than anything the caller can influence like a header).

**Comment by maxiplay** [comment]
Yes backend resources would be used both by a browser and other services. But this rest end point  is specifically called by the feign client. Not the browser. But crsrf protection and authentication are shared for all end points and I want to protect this specific end point from browser.

When you say "split into two populations" you say lauch a new specific service apart to be used by feign client ? and protected by browser but not feign ?

**Comment by dsyer** [comment]
If this endpoint is called only by the Feign client then you can secure it independently (and switch off CSRF). Just add a new `WebSecurityConfigurer` with a low order and a request matcher that only matches this resource.

**Comment by maxiplay** [comment]
Ok it's seems good to me. But I don't see how can I protect the end point from user while authorizing access to my feign client. Do you have an example ? My feign client use zuul and eureka like users to discover services.

**Comment by dsyer** [comment]
How is your Feign client going to authenticate? That's how you protect the endpoint from users (make it something only the Feign client knows or has). Pretty normal security configuration really.

**Comment by maxiplay** [comment]
Now you say, it seems simple. I will try tonight. I'll let you know

**Comment by maxiplay** [comment]
I used requestInterceptor on my feign client to to pass login/password.

    @Bean
	public BasicAuthRequestInterceptor basicAuthRequestInterceptor() {
		return new BasicAuthRequestInterceptor("login", "password");
	}

And I deactivate csrf on the service called. I realized that csrf protection on the gateway is sufficent. I thought zuul proxy forwarded request before csrf protection work.I thought it was necessary to have csrf protection on all services. But in fact no :).

I have shared a module to authenticate a user through my services and now each service is able to log in a user with dedicated database. Not really "micro-service warranty" but not a problem for me.  So now my feign client has its own user and password and I can manage right easily with Spring security. The feign client is considered as a user with specific rights.

Thank you for support. More I use Spring and more I want to use Spring :+1:
